core/src/main/scala/akka/persistence/cassandra/journal/CassandraJournal.scala

```diff
@@ -396,38 +395,64 @@ class CassandraJournal(cfg: Config) extends AsyncWriteJournal
-        row.getInt("ser_id"),
-        row.getString("ser_manifest")
-      ).get
-
-      if (hasMetaColumns(row)) {
-        row.getBytes("meta") match {
-          case null =>
-            event // no meta data
-          case metaBytes =>
-            // has meta data, wrap in EventWithMetaData
-            val metaSerId = row.getInt("meta_ser_id")
-            val metaSerManifest = row.getString("meta_ser_manifest")
-            val meta = serialization.deserialize(
-              Bytes.getArray(metaBytes),
-              metaSerId,
-              metaSerManifest
-            ) match {
-                case Success(m) => m
-                case Failure(_) =>
-                  // don't fail replay/query because of deserialization problem with meta data
-                  // see motivation in UnknownMetaData
-                  UnknownMetaData(metaSerId, metaSerManifest)
-              }
-            EventWithMetaData(event, meta)
+    def deserializeEvent(row: Row)(implicit ec: ExecutionContext): Future[Any] = try {
+
+      def meta: OptionVal[AnyRef] = {
+        if (hasMetaColumns(row)) {
+          row.getBytes("meta") match {
+            case null =>
+              OptionVal.None // no meta data
+            case metaBytes =>
+              // has meta data, wrap in EventWithMetaData
+              val metaSerId = row.getInt("meta_ser_id")
+              val metaSerManifest = row.getString("meta_ser_manifest")
+              val meta = serialization.deserialize(
+                Bytes.getArray(metaBytes),
+                metaSerId,
+                metaSerManifest
+              ) match {
+                  case Success(m) => m
+                  case Failure(_) =>
+                    // don't fail replay/query because of deserialization problem with meta data
+                    // see motivation in UnknownMetaData
+                    UnknownMetaData(metaSerId, metaSerManifest)
+                }
+              OptionVal.Some(meta)
+          }
+        } else {
+          // for backwards compatibility, when table was not altered, meta columns not added
+          OptionVal.None // no meta data
         }
-      } else {
-        // for backwards compatibility, when table was not altered, meta columns not added
-        event // no meta data
       }
+
+      val bytes = Bytes.getArray(row.getBytes("event"))
+      val serId = row.getInt("ser_id")
+      val manifest = row.getString("ser_manifest")
+
+      serialization.serializerByIdentity.get(serId) match {
+        case Some(asyncSerializer: AsyncSerializer) =>
+          Serialization.withTransportInformation(system.asInstanceOf[ExtendedActorSystem]) { () =>
+            asyncSerializer.fromBinaryAsync(bytes, manifest).map { event =>
+              meta match {
+                case OptionVal.None    => event
+                case OptionVal.Some(m) => EventWithMetaData(event, m)
+              }
+            }
+          }
+
+        case _ =>
+          Future {
```

Comment:
Any reason to dispatch this rather than Future.successful it?

Reply:
We had that previously also, otherwise it would run on the actor's thread. https://github.com/akka/akka-persistence-cassandra/pull/353/files#diff-0c1d6befdb42f3861175a7eca0357026L173

Reply:
:+1:

Reply:
You were right, I was thinking about serialization when replying, this is for deserialization. When using mapAsync(1) it would be better to use Future.successful, but with mapAsync(> 1) it would be better to spawn Future. I have added that in additional commit.